azure-cli: immutable command index & add extensions

katexochen opened the pull request:
## Description of changes

Closes https://github.com/NixOS/nixpkgs/issues/272485

First, this PR moves the `commandIndex.json` and `versionCheck.json` from the config dir into the derivation to make them immutable. Manipulating these json files alters the behavior of the CLI which isn't wanted for a nix pkg. It also has weird effects when installing the azure-cli on a non-NixOS system that also has the azure-cli installed from the host package manager and those files already laying around, as both tools will read and these files and might clash, especially when one uses extensions.

Second, this PR enables installing the azure-cli with extensions. Azure extensions are python wheels downloaded from various endpoints. There is a constant endpoint to request the index of all extensions (in all versions). I wrote a script to query and parse this index to find the latest extension version that is compatible with the azure-cli version. The tool can also output nix. A set of extension is added in this PR, too. These are all extensions that do not have any additional dependencies/requirements. The other extensions need to be packaged manually.
I still need to find a convenient way to automatically update the extensions, likely by recreating the extension file.

### Things left to do:

- [x] Recurse into extension attrset on evaluation (I'd love to only expose the extensions via passthru, but it doesn't seem to be possible, as `recurseIntoAttrs` will stop when encountering a derivation. I've added `azure-cli-extensions` for now.)
- [ ] ~~Automatic extension updates~~

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

katexochen committed:
azure-cli: immutable command index

katexochen committed:
azure-cli: add extensions

AkechiShiro commented:
Thanks a lot for your contribution @katexochen, I was wondering who should maintain those extension up to date because they're like mini python pkgs.

So far as you've mentioned there are 2 types of extension :
- Those that can be packaged easily (don't need much dependencies)
- Those that need to be packaged manually.

There is probably a way to automate the upgrade for the first one, for now, running the query/update script manually is already a step in the right direction as we previously had no extension support.

I will try and do some testing of some extension and report how it went.

katexochen commented:
@AkechiShiro there is no automation for updating the `azure-cli` including its dependencies, is there?

AkechiShiro commented:
It's a bit complicated but short answer is no.

If we didn't pin the packages in python-package.nix then we'd fallback on most python modules being automatically upgraded by the bot ryantm but it's best to pin them to endure the builds always works.

I've been working on generating the pinned packages from the tag of the requirements for every azure-cli but I haven't finished yet.

drupol commented on pkgs/tools/admin/azure-cli/default.nix:
```diff
@@ -17,6 +21,47 @@ let
+
+  extensions = import ./extensions.nix { inherit mkAzExtension; };
+
+  extensionDir = stdenv.mkDerivation {
```
I would use `stdenvNoCC` since no compilation is required here.

drupol commented on pkgs/tools/admin/azure-cli/python-packages.nix:
```diff
@@ -30,6 +30,10 @@ let
         sourceRoot = "${src.name}/src/azure-cli-core";
 
+        patches = [
+          ./0001-optional-immutable-configuration-dir.patch
```
Each patch in nixpkgs applied to the upstream source should be documented. Out-of-tree patches, fetched using `fetchpatch`, are preferred.

In each patch comment, please explain the purpose of the patch and link to the relevant upstream issue if possible. If the patch has been merged upstream but is not yet part of the released version, please note the version number or date in the comment such that a future maintainer updating the nix expression will know whether the patch has been incorporated upstream and can thus be removed from nixpkgs.

Furthermore, please use a _stable_ URL for the patch. Rather than, for example, linking to a GitHub pull request of the form `https://github.com/owner/repo/pull/pr_number.patch`, which would change every time a commit is added or the PR is force-pushed, link to a specific commit patch in the form `https://github.com/owner/repo/commit/sha.patch`.

Here are two good examples of patch comments:

```nix
mkDerivation {
  …

  patches = [
    # Ensure RStudio compiles against R 4.0.0.
    # Should be removed next 1.2.X RStudio update or possibly 1.3.X.
    (fetchpatch {
      url = "https://github.com/rstudio/rstudio/commit/3fb2397c2f208bb8ace0bbaf269481ccb96b5b20.patch";
      sha256 = "0qpgjy6aash0fc0xbns42cwpj3nsw49nkbzwyq8az01xwg81g0f3";
    })
  ];
}
```

```nix
mkDerivation {
  …

  patches = [
    # Allow building with bison 3.7
    # PR at https://github.com/GoldenCheetah/GoldenCheetah/pull/3590
    (fetchpatch {
      url = "https://github.com/GoldenCheetah/GoldenCheetah/commit/e1f42f8b3340eb4695ad73be764332e75b7bce90.patch";
      sha256 = "1h0y9vfji5jngqcpzxna5nnawxs77i1lrj44w8a72j0ah0sznivb";
    })
  ];
}
```

katexochen replied:
The patch itself contains a detailed description of what is changed. I've added a comment why we need the patch and about the upstream situation.

drupol commented on pkgs/top-level/all-packages.nix:
```diff
@@ -3440,6 +3440,8 @@ with pkgs;
 
   azure-cli = callPackage ../tools/admin/azure-cli { };
 
+  azure-cli-extensions = recurseIntoAttrs azure-cli.extensions;
```
Shouldn't it use `pkgs/by-name` now ?

katexochen replied:
You mean `azure-cli` in general should be in `pkgs/by-name`, or `azure-cli-extensions` specifically? I'm not sure how you would archive the latter one without moving all of azure-cli into by-name, and the extensions would still need an entry in all-packages.nix, wouldn't they?

Thanks for your review. :)

drupol reviewed (changes requested):
Hi !

Thanks for your contribution.

I've left some comments, let me know if this is clear enough.

malt3 commented on pkgs/tools/admin/azure-cli/extensions.nix:
```diff
@@ -0,0 +1,982 @@
+    pname = "baremetal-infrastructure";
+    version = "2.0.0";
+    url = "https://github.com/Azure/azure-baremetalinfrastructure-cli-extension/releases/download/${version}/baremetal_infrastructure-2.0.0-py2.py3-none-any.whl";
+    sha256 = "8672d29ce35b01141c2acfb9d2be5fedc78a9937d2abc9813e95ec061343264b";
```
```suggestion
    sha256 = "158c4e81eb6d773f523e76813f4cf6958fb0c53d08533c3dccc81edd73e593a0";
```

Azure changed this artifact last week

katexochen replied:
Thanks for the notice, fixed it.

malt3 commented:
Result of `nixpkgs-review pr 276695` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>280 packages built:</summary>
  <ul>
    <li>azure-cli</li>
    <li>azure-cli-extensions.account</li>
    <li>azure-cli-extensions.account.dist</li>
    <li>azure-cli-extensions.acrquery</li>
    <li>azure-cli-extensions.acrquery.dist</li>
    <li>azure-cli-extensions.acrtransfer</li>
    <li>azure-cli-extensions.acrtransfer.dist</li>
    <li>azure-cli-extensions.ad</li>
    <li>azure-cli-extensions.ad.dist</li>
    <li>azure-cli-extensions.adp</li>
    <li>azure-cli-extensions.adp.dist</li>
    <li>azure-cli-extensions.aem</li>
    <li>azure-cli-extensions.aem.dist</li>
    <li>azure-cli-extensions.ai-examples</li>
    <li>azure-cli-extensions.ai-examples.dist</li>
    <li>azure-cli-extensions.aks-preview</li>
    <li>azure-cli-extensions.aks-preview.dist</li>
    <li>azure-cli-extensions.akshybrid</li>
    <li>azure-cli-extensions.akshybrid.dist</li>
    <li>azure-cli-extensions.alb</li>
    <li>azure-cli-extensions.alb.dist</li>
    <li>azure-cli-extensions.alertsmanagement</li>
    <li>azure-cli-extensions.alertsmanagement.dist</li>
    <li>azure-cli-extensions.amg</li>
    <li>azure-cli-extensions.amg.dist</li>
    <li>azure-cli-extensions.amlfs</li>
    <li>azure-cli-extensions.amlfs.dist</li>
    <li>azure-cli-extensions.apic-extension</li>
    <li>azure-cli-extensions.apic-extension.dist</li>
    <li>azure-cli-extensions.appservice-kube</li>
    <li>azure-cli-extensions.appservice-kube.dist</li>
    <li>azure-cli-extensions.authV2</li>
    <li>azure-cli-extensions.authV2.dist</li>
    <li>azure-cli-extensions.automanage</li>
    <li>azure-cli-extensions.automanage.dist</li>
    <li>azure-cli-extensions.automation</li>
    <li>azure-cli-extensions.automation.dist</li>
    <li>azure-cli-extensions.azure-firewall</li>
    <li>azure-cli-extensions.azure-firewall.dist</li>
    <li>azure-cli-extensions.azurelargeinstance</li>
    <li>azure-cli-extensions.azurelargeinstance.dist</li>
    <li>azure-cli-extensions.azurestackhci</li>
    <li>azure-cli-extensions.azurestackhci.dist</li>
    <li>azure-cli-extensions.baremetal-infrastructure</li>
    <li>azure-cli-extensions.baremetal-infrastructure.dist</li>
    <li>azure-cli-extensions.bastion</li>
    <li>azure-cli-extensions.bastion.dist</li>
    <li>azure-cli-extensions.billing-benefits</li>
    <li>azure-cli-extensions.billing-benefits.dist</li>
    <li>azure-cli-extensions.blockchain</li>
    <li>azure-cli-extensions.blockchain.dist</li>
    <li>azure-cli-extensions.blueprint</li>
    <li>azure-cli-extensions.blueprint.dist</li>
    <li>azure-cli-extensions.change-analysis</li>
    <li>azure-cli-extensions.change-analysis.dist</li>
    <li>azure-cli-extensions.cli-translator</li>
    <li>azure-cli-extensions.cli-translator.dist</li>
    <li>azure-cli-extensions.confidentialledger</li>
    <li>azure-cli-extensions.confidentialledger.dist</li>
    <li>azure-cli-extensions.confluent</li>
    <li>azure-cli-extensions.confluent.dist</li>
    <li>azure-cli-extensions.connectedmachine</li>
    <li>azure-cli-extensions.connectedmachine.dist</li>
    <li>azure-cli-extensions.connectedvmware</li>
    <li>azure-cli-extensions.connectedvmware.dist</li>
    <li>azure-cli-extensions.connection-monitor-preview</li>
    <li>azure-cli-extensions.connection-monitor-preview.dist</li>
    <li>azure-cli-extensions.cosmosdb-preview</li>
    <li>azure-cli-extensions.cosmosdb-preview.dist</li>
    <li>azure-cli-extensions.costmanagement</li>
    <li>azure-cli-extensions.costmanagement.dist</li>
    <li>azure-cli-extensions.csvmware</li>
    <li>azure-cli-extensions.csvmware.dist</li>
    <li>azure-cli-extensions.custom-providers</li>
    <li>azure-cli-extensions.custom-providers.dist</li>
    <li>azure-cli-extensions.customlocation</li>
    <li>azure-cli-extensions.customlocation.dist</li>
    <li>azure-cli-extensions.databox</li>
    <li>azure-cli-extensions.databox.dist</li>
    <li>azure-cli-extensions.databricks</li>
    <li>azure-cli-extensions.databricks.dist</li>
    <li>azure-cli-extensions.datadog</li>
    <li>azure-cli-extensions.datadog.dist</li>
    <li>azure-cli-extensions.datafactory</li>
    <li>azure-cli-extensions.datafactory.dist</li>
    <li>azure-cli-extensions.datamigration</li>
    <li>azure-cli-extensions.datamigration.dist</li>
    <li>azure-cli-extensions.dataprotection</li>
    <li>azure-cli-extensions.dataprotection.dist</li>
    <li>azure-cli-extensions.datashare</li>
    <li>azure-cli-extensions.datashare.dist</li>
    <li>azure-cli-extensions.deploy-to-azure</li>
    <li>azure-cli-extensions.deploy-to-azure.dist</li>
    <li>azure-cli-extensions.desktopvirtualization</li>
    <li>azure-cli-extensions.desktopvirtualization.dist</li>
    <li>azure-cli-extensions.dev-spaces</li>
    <li>azure-cli-extensions.dev-spaces.dist</li>
    <li>azure-cli-extensions.devcenter</li>
    <li>azure-cli-extensions.devcenter.dist</li>
    <li>azure-cli-extensions.diskpool</li>
    <li>azure-cli-extensions.diskpool.dist</li>
    <li>azure-cli-extensions.dms-preview</li>
    <li>azure-cli-extensions.dms-preview.dist</li>
    <li>azure-cli-extensions.dnc</li>
    <li>azure-cli-extensions.dnc.dist</li>
    <li>azure-cli-extensions.dns-resolver</li>
    <li>azure-cli-extensions.dns-resolver.dist</li>
    <li>azure-cli-extensions.dynatrace</li>
    <li>azure-cli-extensions.dynatrace.dist</li>
    <li>azure-cli-extensions.edgeorder</li>
    <li>azure-cli-extensions.edgeorder.dist</li>
    <li>azure-cli-extensions.elastic</li>
    <li>azure-cli-extensions.elastic-san</li>
    <li>azure-cli-extensions.elastic-san.dist</li>
    <li>azure-cli-extensions.elastic.dist</li>
    <li>azure-cli-extensions.eventgrid</li>
    <li>azure-cli-extensions.eventgrid.dist</li>
    <li>azure-cli-extensions.express-route-cross-connection</li>
    <li>azure-cli-extensions.express-route-cross-connection.dist</li>
    <li>azure-cli-extensions.fleet</li>
    <li>azure-cli-extensions.fleet.dist</li>
    <li>azure-cli-extensions.fluid-relay</li>
    <li>azure-cli-extensions.fluid-relay.dist</li>
    <li>azure-cli-extensions.footprint</li>
    <li>azure-cli-extensions.footprint.dist</li>
    <li>azure-cli-extensions.front-door</li>
    <li>azure-cli-extensions.front-door.dist</li>
    <li>azure-cli-extensions.fzf</li>
    <li>azure-cli-extensions.fzf.dist</li>
    <li>azure-cli-extensions.graphservices</li>
    <li>azure-cli-extensions.graphservices.dist</li>
    <li>azure-cli-extensions.guestconfig</li>
    <li>azure-cli-extensions.guestconfig.dist</li>
    <li>azure-cli-extensions.hack</li>
    <li>azure-cli-extensions.hack.dist</li>
    <li>azure-cli-extensions.hardware-security-modules</li>
    <li>azure-cli-extensions.hardware-security-modules.dist</li>
    <li>azure-cli-extensions.hdinsightonaks</li>
    <li>azure-cli-extensions.hdinsightonaks.dist</li>
    <li>azure-cli-extensions.healthbot</li>
    <li>azure-cli-extensions.healthbot.dist</li>
    <li>azure-cli-extensions.healthcareapis</li>
    <li>azure-cli-extensions.healthcareapis.dist</li>
    <li>azure-cli-extensions.hpc-cache</li>
    <li>azure-cli-extensions.hpc-cache.dist</li>
    <li>azure-cli-extensions.image-copy-extension</li>
    <li>azure-cli-extensions.image-copy-extension.dist</li>
    <li>azure-cli-extensions.image-gallery</li>
    <li>azure-cli-extensions.image-gallery.dist</li>
    <li>azure-cli-extensions.import-export</li>
    <li>azure-cli-extensions.import-export.dist</li>
    <li>azure-cli-extensions.init</li>
    <li>azure-cli-extensions.init.dist</li>
    <li>azure-cli-extensions.internet-analyzer</li>
    <li>azure-cli-extensions.internet-analyzer.dist</li>
    <li>azure-cli-extensions.ip-group</li>
    <li>azure-cli-extensions.ip-group.dist</li>
    <li>azure-cli-extensions.k8s-extension</li>
    <li>azure-cli-extensions.k8s-extension.dist</li>
    <li>azure-cli-extensions.kusto</li>
    <li>azure-cli-extensions.kusto.dist</li>
    <li>azure-cli-extensions.log-analytics</li>
    <li>azure-cli-extensions.log-analytics-solution</li>
    <li>azure-cli-extensions.log-analytics-solution.dist</li>
    <li>azure-cli-extensions.log-analytics.dist</li>
    <li>azure-cli-extensions.logic</li>
    <li>azure-cli-extensions.logic.dist</li>
    <li>azure-cli-extensions.logz</li>
    <li>azure-cli-extensions.logz.dist</li>
    <li>azure-cli-extensions.maintenance</li>
    <li>azure-cli-extensions.maintenance.dist</li>
    <li>azure-cli-extensions.managedccfs</li>
    <li>azure-cli-extensions.managedccfs.dist</li>
    <li>azure-cli-extensions.managednetworkfabric</li>
    <li>azure-cli-extensions.managednetworkfabric.dist</li>
    <li>azure-cli-extensions.managementpartner</li>
    <li>azure-cli-extensions.managementpartner.dist</li>
    <li>azure-cli-extensions.mixed-reality</li>
    <li>azure-cli-extensions.mixed-reality.dist</li>
    <li>azure-cli-extensions.mobile-network</li>
    <li>azure-cli-extensions.mobile-network.dist</li>
    <li>azure-cli-extensions.monitor-control-service</li>
    <li>azure-cli-extensions.monitor-control-service.dist</li>
    <li>azure-cli-extensions.network-analytics</li>
    <li>azure-cli-extensions.network-analytics.dist</li>
    <li>azure-cli-extensions.networkcloud</li>
    <li>azure-cli-extensions.networkcloud.dist</li>
    <li>azure-cli-extensions.new-relic</li>
    <li>azure-cli-extensions.new-relic.dist</li>
    <li>azure-cli-extensions.next</li>
    <li>azure-cli-extensions.next.dist</li>
    <li>azure-cli-extensions.nginx</li>
    <li>azure-cli-extensions.nginx.dist</li>
    <li>azure-cli-extensions.notification-hub</li>
    <li>azure-cli-extensions.notification-hub.dist</li>
    <li>azure-cli-extensions.nsp</li>
    <li>azure-cli-extensions.nsp.dist</li>
    <li>azure-cli-extensions.offazure</li>
    <li>azure-cli-extensions.offazure.dist</li>
    <li>azure-cli-extensions.orbital</li>
    <li>azure-cli-extensions.orbital.dist</li>
    <li>azure-cli-extensions.palo-alto-networks</li>
    <li>azure-cli-extensions.palo-alto-networks.dist</li>
    <li>azure-cli-extensions.peering</li>
    <li>azure-cli-extensions.peering.dist</li>
    <li>azure-cli-extensions.portal</li>
    <li>azure-cli-extensions.portal.dist</li>
    <li>azure-cli-extensions.powerbidedicated</li>
    <li>azure-cli-extensions.powerbidedicated.dist</li>
    <li>azure-cli-extensions.providerhub</li>
    <li>azure-cli-extensions.providerhub.dist</li>
    <li>azure-cli-extensions.purview</li>
    <li>azure-cli-extensions.purview.dist</li>
    <li>azure-cli-extensions.qumulo</li>
    <li>azure-cli-extensions.qumulo.dist</li>
    <li>azure-cli-extensions.quota</li>
    <li>azure-cli-extensions.quota.dist</li>
    <li>azure-cli-extensions.redisenterprise</li>
    <li>azure-cli-extensions.redisenterprise.dist</li>
    <li>azure-cli-extensions.reservation</li>
    <li>azure-cli-extensions.reservation.dist</li>
    <li>azure-cli-extensions.resource-graph</li>
    <li>azure-cli-extensions.resource-graph.dist</li>
    <li>azure-cli-extensions.resource-mover</li>
    <li>azure-cli-extensions.resource-mover.dist</li>
    <li>azure-cli-extensions.sap-hana</li>
    <li>azure-cli-extensions.sap-hana.dist</li>
    <li>azure-cli-extensions.scenario-guide</li>
    <li>azure-cli-extensions.scenario-guide.dist</li>
    <li>azure-cli-extensions.scheduled-query</li>
    <li>azure-cli-extensions.scheduled-query.dist</li>
    <li>azure-cli-extensions.scvmm</li>
    <li>azure-cli-extensions.scvmm.dist</li>
    <li>azure-cli-extensions.self-help</li>
    <li>azure-cli-extensions.self-help.dist</li>
    <li>azure-cli-extensions.sentinel</li>
    <li>azure-cli-extensions.sentinel.dist</li>
    <li>azure-cli-extensions.site-recovery</li>
    <li>azure-cli-extensions.site-recovery.dist</li>
    <li>azure-cli-extensions.spring</li>
    <li>azure-cli-extensions.spring-cloud</li>
    <li>azure-cli-extensions.spring-cloud.dist</li>
    <li>azure-cli-extensions.spring.dist</li>
    <li>azure-cli-extensions.stack-hci</li>
    <li>azure-cli-extensions.stack-hci-vm</li>
    <li>azure-cli-extensions.stack-hci-vm.dist</li>
    <li>azure-cli-extensions.stack-hci.dist</li>
    <li>azure-cli-extensions.staticwebapp</li>
    <li>azure-cli-extensions.staticwebapp.dist</li>
    <li>azure-cli-extensions.storage-blob-preview</li>
    <li>azure-cli-extensions.storage-blob-preview.dist</li>
    <li>azure-cli-extensions.storage-mover</li>
    <li>azure-cli-extensions.storage-mover.dist</li>
    <li>azure-cli-extensions.storagesync</li>
    <li>azure-cli-extensions.storagesync.dist</li>
    <li>azure-cli-extensions.stream-analytics</li>
    <li>azure-cli-extensions.stream-analytics.dist</li>
    <li>azure-cli-extensions.subscription</li>
    <li>azure-cli-extensions.subscription.dist</li>
    <li>azure-cli-extensions.support</li>
    <li>azure-cli-extensions.support.dist</li>
    <li>azure-cli-extensions.timeseriesinsights</li>
    <li>azure-cli-extensions.timeseriesinsights.dist</li>
    <li>azure-cli-extensions.traffic-collector</li>
    <li>azure-cli-extensions.traffic-collector.dist</li>
    <li>azure-cli-extensions.virtual-network-manager</li>
    <li>azure-cli-extensions.virtual-network-manager.dist</li>
    <li>azure-cli-extensions.virtual-network-tap</li>
    <li>azure-cli-extensions.virtual-network-tap.dist</li>
    <li>azure-cli-extensions.virtual-wan</li>
    <li>azure-cli-extensions.virtual-wan.dist</li>
    <li>azure-cli-extensions.vm-repair</li>
    <li>azure-cli-extensions.vm-repair.dist</li>
    <li>azure-cli-extensions.vmware</li>
    <li>azure-cli-extensions.vmware.dist</li>
    <li>azure-cli-extensions.webapp</li>
    <li>azure-cli-extensions.webapp.dist</li>
    <li>azure-cli-extensions.workloads</li>
    <li>azure-cli-extensions.workloads.dist</li>
    <li>azure-cli.dist</li>
  </ul>
</details>

malt3 reviewed (approved):
I tested this by building a CLI with a select number of extensions and checking if the newly added commands work. LGTM.

katexochen committed:
azure-cli: add passthru function to override extensions

Some syntactical sugar for easy overrides.

katexochen commented:
- Rebased this PR to pull in the v2.56.0 update and updated the extensions accordingly.
- Fixed a minor bug in the jq update script.
- Added a toggle `withImmutableConfig` to disable the immutable config patch when needed.
  Can be built with 
  ```sh
  nix build --impure --expr 'with (import ./. {}); azure-cli.override { withImmutableConfig = false; }'
  ```
- Added `withExtensions` function as syntactical sugar to passthru. CLI with extensions can now be built with
  ```sh
   nix build --impure --expr 'with (import ./. {}); azure-cli.withExtensions [ azure-cli.extensions.aks-preview ]'
   ```

I will create a followup PR with an extensions that needs manual packaging so the structure for this is also given.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3419

AkechiShiro commented:
Thanks a lot for your contribution @katexochen, as well as @malt3, @drupol for your reviews.

I did some quick testing :  
- built with `nixpkgs-review`, 280 packages all built successfully on my side as well.
- Tested building with an extension
- Tested building with `immutableConfig false`

It looks good to me, however I believe there is still some configuration file that are not immutable, not sure if that's something that should be notified upstream as well in https://github.com/Azure/azure-cli/issues/28093, just noting down this behavior here, if anyone wants to take a stab at it, in following PRs, this would require extending the current patch for the immutable config.

<summary>
<details>
<pre lang=bash>
[nix-shell:~/.cache/nixpkgs-review/pr-276695-1/nixpkgs/result]$ ./bin/az configure

Welcome to the Azure CLI! This command will guide you through logging in and setting some default values.

Your settings can be found at /home/zwei/.azure/config
Your current configuration is as follows:

[cloud]
name = AzureCloud

[core]
output = yamlc
collect_telemetry = no
cache_ttl = 10

[logging]
enable_log_file = no

Environment variables:
AZURE_IMMUTABLE_DIR = /nix/store/gb8sqs1r9r9ggwni0sacmiqhw03zlbvh-python3.11-azure-cli-2.56.0/etc/azure
AZURE_EXTENSION_DIR = /nix/store/xbvfvl2xccyvk330pqpqy6dc21hf17bc-azure-cli-extensions

Do you wish to change your settings? (y/N): y   

What default output format would you like?
 [1] json - JSON formatted output that most closely matches API responses.
 [2] jsonc - Colored JSON formatted output that most closely matches API responses.
 [3] table - Human-readable output format.
 [4] tsv - Tab- and Newline-delimited. Great for GREP, AWK, etc.
 [5] yaml - YAML formatted output. An alternative to JSON. Great for configuration files.
 [6] yamlc - Colored YAML formatted output. An alternative to JSON. Great for configuration files.
 [7] none - No output, except for errors and warnings.
Please enter a choice [Default choice(6)]: 2  

Would you like to enable logging to file? (y/N): y

...
...

[nix-shell:~/.cache/nixpkgs-review/pr-276695-1/nixpkgs/result]$ cat ~/.azure/config 
[cloud]
name = AzureCloud

[core]
output = jsonc
collect_telemetry = no
cache_ttl = 10

[logging]
enable_log_file = yes
</pre>
</details>
</summary>

There is one last issue about using `pkgs/by-name` that needs to be addressed, I believe, @drupol could you please let us know, if the move to `pkgs/by-name` needs to be done for the `azure-cli` package : https://github.com/NixOS/nixpkgs/pull/276695#discussion_r1464704089

The rest of the dependencies will still have to stay in python packages, I think.

katexochen commented:
@AkechiShiro Thanks for your review. 

> I believe there is still some configuration file that are not immutable

I think there are two kind of "configuration" files here: The ones that this PR moves into the store aren't real configuration files in the closer sense. They aren't meant for a human to _configure_ the azure-cli, but rather some mechanism to allow dynamic changes to the logic of the program, that's why we would want them to be in the nix store.

`.azure/config`, `.azure/azureProfile.json` are actual configuration, and I think these should stay outside of the store so users can actually set configuration. Users who want that part immutable as well could use home-manager for example.

AkechiShiro commented:
Thanks for mentioning this important distinction on the config files, I do agree with you, it seems to be better to seperate the *human editable* ones from the tool internal one.

And `home-manager` is most likely a better tool to handle the *human* editable one.

AkechiShiro commented:
I'm not sure if it is relevant @katexochen, but there may need to also document this change in the package for previous users in the manual for 24.11, as this change could break some existing behavior if by default the CLI is installed with `immutableConfig = true`, what do you think regarding this ?  
> [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
 (Package updates) Added a release notes entry if the change is major or breaking

katexochen commented:
> but there may need to also document this change in the package for previous users in the manual

Thanks for the reminder, I've added it to the release notes.

SuperSandro2000 commented on pkgs/tools/admin/azure-cli/default.nix:
```diff
@@ -16,7 +28,48 @@ let
+
+  extensionDir = stdenvNoCC.mkDerivation {
+    name = "azure-cli-extensions";
+    phases = [ "installPhase" ];
```
We wouldn't overwrite phases like this. It's there a particular reason they should be disabled?

katexochen replied:
Ture, `dontUnpack = true` should work as well. Or should we just use `runCommand`? We don't really need anything from `mkDerivation` here, do we?

SuperSandro2000 commented on pkgs/tools/admin/azure-cli/default.nix:
```diff
@@ -187,7 +246,12 @@ py.pkgs.toPythonApplication (py.pkgs.buildAzureCliPackage {
+  '' + lib.optionalString (withExtensions != [ ]) ''
+    --set AZURE_EXTENSION_DIR ${extensionDir} \
+  '' + ''
+    --set PYTHONPATH $PYTHONPATH
```
This leaks inputs from all phases, if there any. Better to use the better functions.

katexochen replied:
Yes, this leaks things written in checkPhase into the output, good point. Not sure how to fix it tho.

> Better to use the better functions.

What function?

SuperSandro2000 replied:
Depending on what we need, we can either just prefix python.sitePackages https://github.com/NixOS/nixpkgs/blob/beea7e8ff8b7a3de32b9fe9d7473d99b5c1da503/pkgs/applications/audio/osdlyrics/default.nix#L75 , use withPackages https://github.com/NixOS/nixpkgs/blob/beea7e8ff8b7a3de32b9fe9d7473d99b5c1da503/nixos/modules/services/continuous-integration/buildbot/master.nix#L273 or make pythonPath https://github.com/NixOS/nixpkgs/blob/beea7e8ff8b7a3de32b9fe9d7473d99b5c1da503/nixos/modules/services/misc/tandoor-recipes.nix#L133 

I didn't look into what is more correct and the best solution here though

katexochen replied:
Thanks, I've put this into a separate commit as it was somewhat unreleated to the other changes made.

SuperSandro2000 commented on pkgs/tools/admin/azure-cli/python-packages.nix:
```diff
@@ -30,6 +32,16 @@ let
+          # This is a downstream patch without an commit or PR upstream.
+          # There is an issue to discuss possible solutions upstream:
+          # https://github.com/Azure/azure-cli/issues/28093
+          lib.optional withImmutableConfig ./0001-optional-immutable-configuration-dir.patch
```
Can we make that depend purely on an env?

katexochen replied:
Not sure I get your question. The patch already checks for the env. Are you suggesting to just always apply it?

SuperSandro2000 replied:
> Are you suggesting to just always apply it?

Yes, I think that would be a lot easier

katexochen replied:
Done. :)

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -3413,6 +3413,8 @@ with pkgs;
 
   azure-cli = callPackage ../tools/admin/azure-cli { };
 
+  azure-cli-extensions = recurseIntoAttrs azure-cli.extensions;
```
Are we calling callPackages for the extensions?

katexochen replied:
Good point, we definitely should. You'd say this is the right place for it?

katexochen replied:
Yeah, seems better in all-packages.nix. This way, we can still use the passthru to build a CLI with all extensions like:
```
 nix build --impure --expr 'with (import ./. {}); azure-cli.withExtensions (lib.attrValues azure-cli.extensions)'
 ```
 which will fail if we use `callPackage` inside `azure-cli/default.nix`, as `callPackage` adds something that hasn't go a pname.

katexochen replied:
Never mind, can't be added in all-packages.nix.

katexochen committed:
azure-cli: explicitly construct PYTHONPATH

This way we don't leak inputs from the build environment into the
runtime environment.

Signed-off-by: Paul Meyer <49727155+katexochen@users.noreply.github.com>

SuperSandro2000 commented on pkgs/tools/admin/azure-cli/default.nix:
```diff
@@ -254,7 +254,7 @@ py.pkgs.toPythonApplication (py.pkgs.buildAzureCliPackage {
     --set AZURE_EXTENSION_DIR ${extensionDir} \
   '' + ''
-    --set PYTHONPATH $PYTHONPATH
+    --set PYTHONPATH "${python3.pkgs.makePythonPath propagatedBuildInputs}:$out/${python3.sitePackages}"
```
We should probably make this a fixup of the original commit

katexochen replied:
I didn't introduce or changed that line in a previous commit, it was only moved (that's why GitHub is showing it as added), so I thought a separate commit for a fix that is otherwise unrelated to this PR would make sense.

johanot commented:
I'd really like to see this merged as well. @SuperSandro2000 @AkechiShiro what do you see missing to make this happen?

jonringer commented:
can't merge until @drupol re-reviews

pogopaule commented:
Not sure if I should open a new issue or comment here.
Using nixpkgs-master with devenv.sh on a Mac I now get this error for azure-cli

```
error: builder for '/nix/store/386r8a40slb8iyk31q2qikmdaav3wn5j-python3.11-azure-cli-2.58.0.drv' failed with exit code 1;
       last 10 log lines:
       >   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
       >   File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
       >   File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
       >   File "<frozen importlib._bootstrap_external>", line 940, in exec_module
       >   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
       >   File "/nix/store/zh5ma5wa7j1vjcln9glpy8as6rplqzi5-python3.11-azure-multiapi-storage-1.2.0/lib/python3.11/site-packages/azure/multiapi/storagev2/blob/__init__.py", line 1, in <module>
       >     __import__('pkg_resources').declare_namespace(__name__)
       >     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
       > ModuleNotFoundError: No module named 'pkg_resources'
       > To check existing issues, please visit: https://github.com/Azure/azure-cli/issues
       For full logs, run 'nix log /nix/store/386r8a40slb8iyk31q2qikmdaav3wn5j-python3.11-azure-cli-2.58.0.drv'.
error: 1 dependencies of derivation '/nix/store/n796k39ds5v45nwc1ramcs7v8hipyjsq-devenv-profile.drv' failed to build
error:
       error: 1 dependencies of derivation '/nix/store/n1n60rpf10dk7p9l4a9x8jyrmvz51qm1-devenv-shell-env.drv' failed to build

✖ Command produced the following output:


Error:   × Command `/nix/store/ydplld92imis3hlpcrxbfj2nnds6zg3f-nix-devenv-2.21.0pre20240315_c5bbf14/bin/nix --show-trace --extra-experimental-features nix-command --extra-experimental-features flakes --option warn-dirty false --option eval-cache false --keep-going --max-jobs 4 print-dev-env --profile /Users/fabian/dev/optinode/optinode_backend/.devenv/gc/shell` failed with with exit
  │ code 1

direnv: loading the environment failed
direnv: error exit status 1


```

katexochen commented:
https://github.com/NixOS/nixpkgs/pull/276695/commits/aacf05daec3141ce2bb34fd7c021a86401ac8c51 cleaned up python path and removed setuptools from the runtime dependencies, which seems to be required now by the `self-check` command which runs in `installCheckPhase`. There was a version bump on master since the last CI run on this PR, that's why it wasn't catched.

mpcusack commented:
@katexochen Could you give an example of how to install `azure-cli` with an extension in a `configuration.nix`?

katexochen commented:
> @katexochen Could you give an example of how to install `azure-cli` with an extension in a `configuration.nix`?

This PR added a release note with an example that should hopefully work:


```nix
environment.systemPackages = [
  (azure-cli.withExtensions [ azure-cli.extensions.aks-preview ]);
];
```

mpcusack-sequelae commented:
Ah hah. Working for me now (without the `;`). Thanks!

JGerardoHarlan commented:
I'm not seeing azure-devops in the extensions - some poking around suggests that's because `query-extension-index.sh` was run with the `--cli-version` flag below `2.30.0`. `az --version` reports 2.58.0 on nixos-unstable (which is how I've found to currently access these changes), Was this done intentionally? Am I just missing something obvious?

 I'm attempting to use universal package artifacts in devops, and this PR was very promising until I realized the devops extension is missing.

katexochen commented:
> I'm not seeing azure-devops in the extensions - some poking around suggests that's because `query-extension-index.sh` was run with the `--cli-version` flag below `2.30.0`. `az --version` reports 2.58.0 on nixos-unstable (which is how I've found to currently access these changes), Was this done intentionally? Am I just missing something obvious?
> 
> I'm attempting to use universal package artifacts in devops, and this PR was very promising until I realized the devops extension is missing.

Hey @JGerardoHarlan, the reason `azure-devops` isn't package yet is not due to the version used during generation but the extension requiring external dependencies:

```json
{
  "pname": "azure-devops",
  "description": "Tools for managing Azure DevOps.",
  "version": "1.0.0",
  "url": "https://github.com/Azure/azure-devops-cli-extension/releases/download/20240206.1/azure_devops-1.0.0-py2.py3-none-any.whl",
  "sha256": "658a2854d8c80f874f9382d421fa45abf6a38d00334737dda006f8dec64cf70a",
  "license": "MIT",
  "requires": [
    "distro (==1.3.0)",
    "distro==1.3.0"
  ]
}
```
This PR only packages dependencies that have no dependencies and thus can be packaged automatically.

I'll try to find the time this weekend to package the extension as an example for how extensions with external dependencies can be added.
